esx_service/utils/vmdk_utils.py

```diff
@@ -63,44 +63,57 @@
 VMDK_RETRY_COUNT = 5
 VMDK_RETRY_SLEEP = 1
 
-def init_datastoreCache():
+# For managing resource locks.
+lockManager = threadutils.LockManager()
+
+
+def init_datastoreCache(force=False):
     """
-    Initializes the datastore cache with the list of datastores accessible from local ESX host.
+    Initializes the datastore cache with the list of datastores accessible
+    from local ESX host. force=True will force it to ignore current cache
+    and force init
     """
     global datastores
     logging.debug("init_datastoreCache: %s", datastores)
 
-    si = vmdk_ops.get_si()
+    with lockManager.get_lock("init_datastoreCache"):
+        if datastores:
+            if force:
+                datastores = None
+            else:
+                return
+
+        si = vmdk_ops.get_si()
 
-    #  We are connected to ESX so childEntity[0] is current DC/Host
-    ds_objects = \
-      si.content.rootFolder.childEntity[0].datastoreFolder.childEntity
-    datastores = []
+        #  We are connected to ESX so childEntity[0] is current DC/Host
+        ds_objects = si.content.rootFolder.childEntity[0].datastoreFolder.childEntity
+        datastores = []
+
+        for datastore in ds_objects:
+            dockvols_path, err = vmdk_ops.get_vol_path(datastore.info.name)
```

Comment:
L89 -92 could be simply:
```python
    tmp_ds = []
    for datastore si.content.rootFolder.childEntity[0].datastoreFolder.childEntity:
        ... same operation, using tmp_ds instead of datastore
   datastores = tmp_ds
```